release/models/network-instance/openconfig-programming-errors.yang

```diff
@@ -0,0 +1,240 @@
+
+          uses ip-routes-common-config;
+        }
+
+        container state {
+          config false;
+          description
+            "IP route programming errors State parameters";
+
+          uses ip-routes-common-config;
+        }
+
+        container failed-routes {
+          config false;
+          description
+            "Surrounding container for the list of routes that fail programming.";
+
+          list failed {
+            key "prefix";
+
+            description
+              "A prefix that was attempted to the programmed into hardware, but the
+              programming operation failed.";
+
+            leaf prefix {
+              type leafref {
+                path "../state/prefix";
+              }
+              description
+                "Reference to the prefix that keys the failed list.";
+            }
+
+            container state {
+              description
+                "Operational state parameters relating to a failed programming
+                operation.";
+              uses route-state-common;
+            }
+          }
+        }
+
+        container stale-routes {
+          config false;
+          description
+            "Surrounding container for the list of routes that are currently in
+            a stale state.";
+
+          list stale {
+            key "prefix";
+
+            description
+              "A prefix that is currently installed in hardware, but a subsequent
+              operation to update its programming failed - such that the entry in
+              hardware is stale.";
+
+            leaf prefix {
+              type leafref {
+                path "../state/prefix";
+              }
+              description
+                "Reference to the prefix that keys the stale list.";
+            }
+
+            container state {
+              description
+                "Operational state parameters relating to a stale route.";
+              uses route-state-common;
+            }
+          }
+        }
+
+        container drop-routes {
+          config false;
+          description
+            "Surrounding container for the list of routes that are currently in
+            a drop state.";
+
+          list drop {
+            key "prefix";
+
+            description
+              "A prefix that is currently installed in hardware but with an explicit
+              instruction that it should discard packets that are destined towards
+              it. This list includes both intentional and un-intentional
+              drop routes.";
+
+            leaf prefix {
+              type leafref {
+                path "../state/prefix";
+              }
+              description
+                "Reference to the prefix that keys the drop list.";
+            }
+
+            container state {
+              description
+                "Operational state parameters relating to a drop route.";
+              uses route-state-common;
+            }
```

Comment:
While I understand the operational use case to detect the root cause of packet loss due to routes pointing to null, but successfully programmed drop routes are not route programming errors I think.    Routes with a null nexthop are already represented in the aft model.    Therefore, I think this container may not be necessary and is duplicative of AFT.

Here is the traversal one would need to do to resolve a route which is pointing to a null interface:
```
/openconfig-network-instance:network-instances/network-instance/afts/ipv4-unicast/ipv4-entry/prefix
/openconfig-network-instance:network-instances/network-instance/afts/next-hop-groups/next-hop-group/id
/openconfig-network-instance:network-instances/network-instance/afts/next-hops/next-hop/interface-ref/state/interface
```

I am not sure what an un-intentional drop route is?    @akalluru1 can you elaborate?

Reply:
un-intentional drops are routes which are programmed as drop routes due to some error. There are cases where we program a drop route if the resolution of the nexthopgroup fails. Another thing to clarify is that the AFT export may not always capture drop routes in the way you indicated. We export drop routes as a prefix pointing to a NHG with no nexthop(s) in it ( an empty nexthopgroup ).

So overall the telemetry data here is trying to capture the case that there is a drop route programmed in the hw, trying to figure whether that is due to an error/at which layer that error is would be better served by querying other parts of the model like the RIB/looking at the config.

Reply:
I've removed the un-intentional/intentional terminology from the latest diff. Please take a look & let us know if something is not clear.